[Tan Jie Wei] iP

jiewei98 opened the pull request:
# **Duke**
> "Believe You Can and You're Halfway There" - Theodore Roosevelt ([Source](https://www.motivateamazebegreat.com/2017/07/believe-you-can-and-youre-halfway-there-meaning.html))
 
Duke is a task reminder application that has the following applications:
- Task creation (can be tagged as a To-do, Event or Deadline)
- Auto-saving of tasks when tasks are updated
- Searching of tasks by keyword

All you need to do is,
1. download it from [here](https://github.com/jiewei98/ip) 
2. run the Duke.main() file
3. and add your tasks!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (To be added)
- [ ] Reminders (To be added)

Here is the `main` method in the Duke.main() file:
```ruby
public static void main(String[] args) {

        Ui ui = new Ui();
        ui.displayGreeting();

        Storage storage = new Storage();
        ArrayList<Task> list = storage.load();

        Parser parser = new Parser();
        CommandList commandList = new CommandList();

        while (!parser.isTerminated) {
            try {
                commandList.execute(parser.readInput(), parser, list, storage);
            } catch (IOException e) {
                System.out.println("Unable to read input!");
            }
        }
    }
```

Thanks for the read! :grin:

Commit:
Add Gradle support

Commit:
Added level 1 functionalities

Commit:
Added level 2 functionalities

Commit:
Added level 3 functionalities

Commit:
Added level 4 functionalities

Commit:
Added Automated Text UI Testing

Commit:
Added level 5 functionalities

Commit:
Added level 6 functionalities

Commit:
Added level 7 functionalities

Commit:
Added level 8 functionalities

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Commit:
Fixed issue when merging branches

Commit:
Revert "Fixed issue when merging branches"

This reverts commit d1fb2275b6eba94dd1dbdca71fbe202872e95e59.

Commit:
Revert "Revert "Fixed issue when merging branches""

This reverts commit 55228b5f2c1531a1ac7d9003dac6c472fd74a40a.

Commit:
Add more OOP

Commit:
Add packages

Commit:
Add basic JUnit test

Commit:
Add JavaDocs

Commit:
Comply to coding standards

Commit:
Add Level 9 functionalities

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

Commit:
Change mainClassName

Graphcalibur commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,115 @@
+            StringBuffer sb = new StringBuffer();
+            String currLine;
+
+            while ((currLine = br.readLine()) != null) {
```
This code block is a little difficult to read. Maybe include some line breaks between each condition?

Graphcalibur commented on src/main/java/Duke/Ui.java:
```diff
@@ -0,0 +1,12 @@
+package Duke;
+
+public class Ui {
```
Does this need to be a class of its own?

Graphcalibur commented on src/main/java/Duke/CommandList.java:
```diff
@@ -0,0 +1,133 @@
+     * @param list Arraylist of Tasks to manipulate
+     * @param storage Storage object to load/store list of Tasks
+     */
+    public void execute(String input, Parser parser, ArrayList<Task> list, Storage storage) {
```
I find this function to be long and cumbersome to read. Perhaps you could include some line breaks, or extract some parts of the code into a separate function?

Graphcalibur commented on src/main/java/Duke/CommandList.java:
```diff
@@ -0,0 +1,133 @@
+            for (int i = 0; i < list.size(); i++) {
+                System.out.println((i + 1) + "." + list.get(i));
+            }
+        } else if (firstWord.equals("mark") || firstWord.equals("unmark")) { // mark/unmark task
```
It seems like it's possible for this part of the code to throw an exception (such as if secondword can't be converted to an int or if index goes outside the bounds of the ArrayList). Perhaps you can catch and handle these exceptions?

Graphcalibur commented on src/main/java/Duke/CommandList.java:
```diff
@@ -0,0 +1,133 @@
+            }
+        } else if (firstWord.equals("find")) { // find tasks by a keyword
+            String secondWord = inputs[1];
+            List<Task> matches = new ArrayList<>();
```
Is there a reason why matches is of type List<> and not ArrayList<>?

Graphcalibur commented on src/main/java/Duke/CommandList.java:
```diff
@@ -0,0 +1,133 @@
+                    parser.terminate();
+                }
+                try {
+                    SimpleDateFormat dateFormatter = new SimpleDateFormat("dd/MM/yyyy HHmm");
```
There's some copy-pasted code here with parsing and formatting the date. Maybe you could extract this out into a function to avoid code duplication?

Graphcalibur reviewed (commented):
Overall, your code mostly seems to follow good practices, but there are a few issues I have that I highlighted in this review.

Commit:
Add basic GUI for duke

Commit:
Add assertions to GUI

Checks that scene object passed to stage during setScene is not null
as a null scene will cause the GUI to not load properly.

Also check for response from CommandList to ensure that response
is not an empty String.

Commit:
Improve code quality and remove unnecessary files

jiewei98 committed:
Merge pull request #2 from jiewei98/branch-A-Assertions

Add assertions to GUI

jiewei98 committed:
Resolve PR conflict for Duke.java

jiewei98 committed:
Merge branch 'branch-A-CodeQuality'

Resolve merge conflict

jiewei98 committed:
Remove TestUi.java as it has been renamed

jiewei98 committed:
Fix typo in getResponse method

jiewei98 committed:
Add update feature to Duke

Able to update Tasks now by specifying index of task as well as new
description together with the date if applicable.

Refactor the code in CommandList to apply SLAP.

jiewei98 committed:
Improve GUI

Add welcome and farewell message from Duke
Fix wrapping and auto truncate issue for long texts

jiewei98 committed:
Add product screenshot and fix clearAll method not updating list issue

jiewei98 committed:
Set theme jekyll-theme-midnight

jiewei98 committed:
Set theme jekyll-theme-slate

jiewei98 committed:
Edit README, add index constraint checks and update Ui screenshot

jiewei98 committed:
Merge branch 'master' of https://github.com/jiewei98/ip

jiewei98 committed:
Update build.gradle path to correct main class
